table/block.h

```diff
@@ -539,18 +541,30 @@ class IndexBlockIter final : public BlockIter<BlockHandle> {
 
   void Invalidate(Status s) { InvalidateBase(s); }
 
+  bool IsValuePinned() const override {
+    return (have_first_key_ && global_seqno_ != kDisableGlobalSequenceNumber)
+               ? false
+               : BlockIter::IsValuePinned();
+  }
+
  private:
   // Key is in InternalKey format
   bool key_includes_seq_;
   bool value_delta_encoded_;
+  bool have_first_key_;  // value includes first_internal_key
   BlockPrefixIndex* prefix_index_;
   // Whether the value is delta encoded. In that case the value is assumed to be
   // BlockHandle. The first value in each restart interval is the full encoded
   // BlockHandle; the restart of encoded size part of the BlockHandle. The
   // offset of delta encoded BlockHandles is computed by adding the size of
   // previous delta encoded values in the same restart interval to the offset of
   // the first value in that restart interval.
-  BlockHandle decoded_value_;
+  IndexValue decoded_value_;
+
+  // First internal key according to current index entry, but with sequence
+  // number overwritten to global_seqno_. Only used if global_seqno_ is enabled.
+  IterKey first_key_;
```

Comment:
`IterKey` is a 65-byte giant object. I'm worried that we always pay this memory overhead even if no first key is provided, which will be majority of the case. An relatively easy way to solve it is to make it a unique_ptr and allocate on-demand. The dynamically allocated data structure can even cover `global_seqno_`, which is another 8 bytes.

Reply:
Oops, yes, I did it but forgot to `git unstash` at some point. Coming in the next update.